Unicode/ustring.c

```diff
@@ -301,33 +301,46 @@ return( NULL );
 
 unichar_t *u_copyn(const unichar_t *pt, long n) {
     unichar_t *res;
+
+    if ( pt && ++n>0 ) {
 #ifdef MEMORY_MASK
-    if ( n*sizeof(unichar_t)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
+	if ( n>MEMORY_MASK/sizeof(unichar_t) )
+	    n = MEMORY_MASK/sizeof(unichar_t);
 #endif
-    res = (unichar_t *) malloc((n+1)*sizeof(unichar_t));
-    memcpy(res,pt,n*sizeof(unichar_t));
-    res[n]='\0';
-return(res);
+	if ( (res=(unichar_t *)(malloc(n*sizeof(unichar_t)))) ) {
+	    if ( (--n) ) memcpy(res,pt,n*sizeof(unichar_t));
+	    res[n] = 0;
+	    return( res );
+	}
+    }
+    return( NULL );
 }
 
 unichar_t *u_copynallocm(const unichar_t *pt, long n, long m) {
+/* Alloc space for m unichar chars and copy n unichar chars */
     unichar_t *res;
+
+    if ( pt && ++n>0 && ++m>0 && m>=n ) {
 #ifdef MEMORY_MASK
-    if ( n*sizeof(unichar_t)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
+	if ( n>MEMORY_MASK/sizeof(unichar_t) )
+	    n = MEMORY_MASK/sizeof(unichar_t);
 #endif
-    res = malloc((m+1)*sizeof(unichar_t));
-    memcpy(res,pt,n*sizeof(unichar_t));
-    res[n]='\0';
-return(res);
+	if ( (res=(unichar_t *)(calloc(m,sizeof(unichar_t)))) ) {
+	    if ( (--n) ) memcpy(res,pt,n*sizeof(unichar_t));
+	    return( res );
+	}
+    }
+    return( NULL );
 }
 
 unichar_t *u_copy(const unichar_t *pt) {
-    if(pt)
-return u_copyn(pt,u_strlen(pt));
+/* Copy unichar_t string and if error, return( "" ). */
+    long n;
+
+    if ( pt && (n=u_strlen(pt))>0 )
+	return( u_copyn(pt,n) );
 
-return((unichar_t *)0);
+    return( calloc(1,sizeof(unichar_t)) );
```

Comment:
How often does this function get called with a NULL pointer? Will returning something to be freed in that case cause leaks?

Reply:
The code was already there.... and I think it was in fontforgeexe/displayfonts.c that got my attention in cleaning-up after doing a printout. It's a lot easier to cleanup a bunch of mallocs.
I think it may be a problem attempting to differentiate a constant (unichar_t *)0 which can't be free, vs an alloc(), which can be free().
Running valgrind tells you there's a lot of stuff that can be cleaned up.